resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -698,10 +701,15 @@ private void finishResolve(Promise<T> promise) {
                 .append(' ');
             }
         }
-        final UnknownHostException cause = new UnknownHostException(buf.toString());
-
-        resolveCache.cache(hostname, additionals, cause, parent.ch.eventLoop());
-        promise.tryFailure(cause);
+        final UnknownHostException unknownHostException = new UnknownHostException(buf.toString());
+        if (cause == null) {
+            // Only cache if the failure was not because of an IO error / timeout that was caused by the query
```

Comment:
couldn't the cache decide if it wants to retain or not instead of always making that decision here?

Reply:
I dont think it will make any sense to put this in the cache ever as it was failed due a timeout or IO error to the dns server (especially an timeout is most likely caused because the dns server enforce some query limit, like public google name servers).

Reply:
@Scottmitch That said I see you point that the cache could handle it.... If we want to make this part of the cache impl I would at least suggest we not cache it in our default impl.  WDYT ?

Reply:
FWIW, I don't think caching transport errors ever really makes much sense: the cache has no particular insight into whether these errors are likely to be transient or permanent, and has to pick some arbitrary TTL to use for this cache entry. Kinda ends up making no sense.

Reply:
@Lukasa +1 ... @Scottmitch you are strong about it ?

Reply:
I think we should either:

- Use the cache API consistently, and let the cache decide (the assumption is the cache already has a policy for this)
- Remove the error path from the caching API completely if we think it is of little use

It doesn't make sense to me to conditionally call the cache API.

Reply:
@Scottmitch removing the error-path from the caching API makes no sense as this would mean we not cache resolve failures that happen because there a no `A` or `AAAA` records present.

Reply:
Agreed. One of the PR goals is state as:

> At the moment there is not way for the user to know if resolving a domain was failed because the domain was unkown or because of an IO error / timeout

How will a user differentiate between the two scenarios? Is the intention that a user will get a failure, and inspect the cause of that failure? Seems like you don't want the cache to use the same technique to w/e take action it may want to (clear existing state, cache failure for some amount of time ... seems to be the current assumption of the API, etc...).

Reply:
The user would check `UnkownHostException.getCause() instanceof DnsNameResolverException`... Like I said I am not super strong on this but honestly I can't think of any good use-case that would make sense to have the cache handle this. But maybe I am wrong... That said if you prefer I can push it to the cache and just make our default implement ignore these. WDYT ?

Reply:
one thing that would be "hard" to do is retries as I did in the second commit here if the cache handles it. As then there is no good way to understand if it makes sense to retry or not as if its in the cache we would basically retry and get the same from the cache all the time

Reply:
> The user would check UnkownHostException.getCause() instanceof DnsNameResolverException

Is it reasonable to provide a utility method which provides this functionality? Propagating the cause exception list can be easy to get wrong. For example what if we later put this exception 2 levels deep instead of at the top level?

> That said if you prefer I can push it to the cache and just make our default implement ignore these.

I'm also not strong, but it seems like the assumption is existing implementations of `DnsCache` are accounting for these scenarios ... unless we think this is an API bug. Either way I think it would be useful to clarify the API. Either saying something like `this won't be called with timeout/cancel/transport exceptions` or `this will be called for timeout/cancel/transport exceptions and if the cache wants to ignore these you should use <utility method> to detect these types of exceptions`.

Reply:
> Is it reasonable to provide a utility method which provides this functionality? Propagating the cause exception list can be easy to get wrong. For example what if we later put this exception 2 levels deep instead of at the top level?

Yep sure I can add such a method. Something like `isCausedByTransportError(...)` or whatever. Any good idea for a name ? 

 > I'm also not strong, but it seems like the assumption is existing implementations of DnsCache are accounting for these scenarios ... unless we think this is an API bug. Either way I think it would be useful to clarify the API. Either saying something like this won't be called with timeout/cancel/transport exceptions or this will be called for timeout/cancel/transport exceptions and if the cache wants to ignore these you should use <utility method> to detect these types of exceptions.

Please see https://github.com/netty/netty/pull/7411#discussion_r152081386

Reply:
> As then there is no good way to understand if it makes sense to retry or not as if its in the cache we would basically retry and get the same from the cache all the time

I would imagine the retry layers and the cache layer are independent. If the retry layer wants to retry due to a transport failure/timeout it is free todo so. If the cache wants to return a result that fails the request immediately it is also free todo so. Do we need to couple the 2 layers?

Also does clarifying the javadocs either way make sense?

Reply:
@Scottmitch what I am saying is that if we want to have the retry logic as part of the `DnsNameResolver` it will only work if the cache does not cache transport errors as otherwise the cache will just return the same result for each retry. 

And yep clarification of the java docs makes sense once we agree and know what to do ;)

Reply:
I'm fine with the change ... the scenarios I'm thinking of are theoretical and in the optimization territory. Since the APIs are experimental we can always change this later if necessary.